app/models/installment.rb

```diff
@@ -129,10 +129,59 @@ class Installment < ApplicationRecord
     end
   }
 
-  scope :missed_for_purchase, -> (purchase) {
-    product_installment_ids = purchase.link.installments.where(seller_id: purchase.seller_id).alive.published.pluck(:id)
-    seller_installment_ids = purchase.seller.installments.alive.published.filter_map do |post|
-      post.id if post.purchase_passes_filters(purchase)
+  scope :seller_or_product_or_variant_type_for_purchase, -> (purchase) {
+    where(
+      "(installment_type IN (?) AND (installments.link_id = ? OR installments.workflow_id IS NOT NULL)) OR installment_type = ?",
+      [PRODUCT_TYPE, VARIANT_TYPE],
+      purchase.link_id,
+      SELLER_TYPE
+    )
+  }
```

Comment:
This is the set of all valid installments for a purchase. 

Overall, the idea is to used this scope to create a base set and filter out as per need, demonstrated in  `missed_for_purchase` and `CustomerService`